klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+
+        ## Drawing polygons here.
+        rule_lay_dt = rule_data_type_map.index(rule_name) + 1
+        if cell is not None:
+            for p in polygons:
+                draw_polygons(p.text, cell, RULE_LAY_NUM, rule_lay_dt, PATH_WIDTH)
+
+        ## Clearing memeory
+        in_item = False
+        elem.clear()
+
+    # Writing final marker gds file
+    if lib is not None:
+        output_gds_path = f'{results_database.replace(".lyrdb", "")}_markers.gds'
+        lib.write_gds(output_gds_path)
+    else:
+        logging.error("Failed to get any results in the lyrdb database.")
+        exit(1)
+
+    # Saving analysis rule deck.
+    for r in rule_data_type_map:
+        rule_lay_dt = rule_data_type_map.index(r) + 1
+        pass_patterns_rule = f'''
+        pass_marker.interacting( text_marker.texts("{r}") ).output("{r}{RULE_STR_SEP}pass_patterns", "{r}{RULE_STR_SEP}pass_patterns polygons")
+        '''
+        fail_patterns_rule = f'''
+        fail_marker2.interacting(fail_marker.interacting(text_marker.texts("{r}")) ).or( fail_marker.interacting(text_marker.texts("{r}")).not_interacting(fail_marker2) ).output("{r}{RULE_STR_SEP}fail_patterns", "{r}{RULE_STR_SEP}fail_patterns polygons")
+        '''
+        false_pos_rule = f'''
+        pass_marker.interacting(text_marker.texts("{r}")).interacting(input({RULE_LAY_NUM}, {rule_lay_dt})).output("{r}{RULE_STR_SEP}false_positive", "{r}{RULE_STR_SEP}false_positive occurred")
+        '''
+        false_neg_rule = f'''
+        ((fail_marker2.interacting(fail_marker.interacting(text_marker.texts("{r}")))).or((fail_marker.interacting(input(11, 222).texts("{r}")).not_interacting(fail_marker2)))).not_interacting(input({RULE_LAY_NUM}, {rule_lay_dt})).output("{r}{RULE_STR_SEP}false_negative", "{r}{RULE_STR_SEP}false_negative occurred")
+        '''
+
+        analysis_rules.append(pass_patterns_rule)
+        analysis_rules.append(fail_patterns_rule)
+        analysis_rules.append(false_pos_rule)
+        analysis_rules.append(false_neg_rule)
+
+    for r in rules_tested:
+        if r in rule_data_type_map:
+            continue
+
+        pass_patterns_rule = f'''
+        pass_marker.interacting( text_marker.texts("{r}") ).output("{r}{RULE_STR_SEP}pass_patterns", "{r}{RULE_STR_SEP}pass_patterns polygons")
+        '''
+        fail_patterns_rule = f'''
+        fail_marker2.interacting(fail_marker.interacting(text_marker.texts("{r}")) ).or( fail_marker.interacting(text_marker.texts("{r}")).not_interacting(fail_marker2) ).output("{r}{RULE_STR_SEP}fail_patterns", "{r}{RULE_STR_SEP}fail_patterns polygons")
+        '''
+
+        analysis_rules.append(pass_patterns_rule)
+        analysis_rules.append(fail_patterns_rule)
+
+    with open(output_runset_path, "w") as runset_analysis:
+        runset_analysis.write("".join(analysis_rules))
+
+    return output_gds_path, output_runset_path
+
+
+def build_tests_dataframe(unit_test_cases_dir, target_table):
```

Comment:
I wonder if could leverage pytest collection customization here? https://docs.pytest.org/en/7.1.x/example/pythoncollection.html and have each test case represented as an actual python test.

Reply:
Let me look at that.

Reply:
One possible way to implement this could be to have a single pytest test case that can run an arbitrary drc rule (similar to what you already have with `run_test_case` in https://github.com/google/globalfoundries-pdk-libs-gf180mcu_fd_pv/blob/e318be650f105e88dd24dc10391dfb0ddc58553d/klayout/drc/testing/run_regression.py#L186) and populate the DRC deck/gds usinga parameterized test data:
- https://docs.pytest.org/en/latest/example/parametrize.html#different-options-for-test-ids
or even better parameterized fixture: 
- https://docs.pytest.org/en/latest/how-to/fixtures.html#parametrizing-fixtures

wdyt?

Reply:
@mithro can also have good input there about making the test methodology fit better with existing python testing best practices.

Reply:
will be addressed #8